ql/src/test/org/apache/hadoop/hive/metastore/TestListPartitionsWithXIncludeParams.java

```diff
@@ -0,0 +1,234 @@
+import org.apache.hadoop.hive.metastore.api.Table;
+import org.apache.hadoop.hive.metastore.client.TestListPartitions;
+import org.apache.hadoop.hive.metastore.client.builder.PartitionBuilder;
+import org.apache.hadoop.hive.metastore.conf.MetastoreConf;
+import org.apache.hadoop.hive.metastore.minihms.AbstractMetaStoreService;
+import org.apache.hadoop.hive.metastore.utils.MetaStoreServerUtils;
+import org.apache.hadoop.hive.metastore.utils.MetaStoreUtils;
+import org.apache.hadoop.hive.ql.exec.SerializationUtilities;
+import org.apache.hadoop.hive.ql.optimizer.ppr.PartitionExpressionForMetastore;
+import org.apache.hadoop.hive.ql.plan.ExprNodeGenericFuncDesc;
+import org.apache.thrift.TException;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import static junit.framework.TestCase.assertTrue;
+import static org.apache.hadoop.hive.metastore.utils.MetaStoreUtils.convertToGetPartitionsByNamesRequest;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+
+/**
+ * Test class for list partitions with configurable include/exclude pattern on parameters.
+ * Embedded Metastore uses JDO, Remote Metastore uses direct SQL to query the partitions:
+ * MetaStoreFactoryForTests.getMetaStores()
+ */
+@RunWith(Parameterized.class)
+@Category(MetastoreCheckinTest.class)
+public class TestListPartitionsWithXIncludeParams
+    extends TestListPartitions {
+
+  private Configuration hiveConf;
+  private Set<String> includeKeys = new HashSet<>();
+  private Set<String> excludeKeys = new HashSet<>();
+  private Map<String, String> partParams = new HashMap<>();
+
+  public static class PartitionExpressionForMetastoreTest extends PartitionExpressionForMetastore {
+    // MetaStoreTestUtils.setConfForStandloneMode will change the default PartitionExpressionForMetastore
+    // to DefaultPartitionExpressionProxy, which doesn't support deserializing the Hive filter from a byte array.
+    // As this test sits inside the hive-exec module, it's safe to set the hive.metastore.expression.proxy to
+    // PartitionExpressionForMetastoreTest.
+  }
+
+  public TestListPartitionsWithXIncludeParams(String name,
+      AbstractMetaStoreService metaStore) {
+    super(name, metaStore);
+    partParams.put("key1", "value1");
+    partParams.put("akey1", "avalue1");
+    partParams.put("akey10", "avalue10");
+    partParams.put("excludekey1", "value1");
+    partParams.put("excludekey2", "value1");
+    includeKeys.add("key1");
+    includeKeys.add("akey1");
+    excludeKeys.add("excludekey1");
+    excludeKeys.add("excludekey2");
+    hiveConf = metaStore.getConf();
+    MetastoreConf.setVar(hiveConf, MetastoreConf.ConfVars.EXPRESSION_PROXY_CLASS,
+        PartitionExpressionForMetastoreTest.class.getName());
+    MetastoreConf.setVar(hiveConf,
+        MetastoreConf.ConfVars.METASTORE_PARTITIONS_PARAMETERS_INCLUDE_PATTERN, "%k_y_");
```

Comment:
Wouldn't this include "excludekey1" and "excludekey2"?

Reply:
Yes, exactly. with `"%k_y_"` alone should include the "excludekey1" and "excludekey2", but we also configure `%exclu%` as the exclude pattern, so both of them would be excluded.

Reply:
I see, so exclude pattern has higher priority compared to include pattern